### PR TITLE
New package: libq

### PR DESCRIPTION
**Pull request by metalMajor**
Specify library name and version:  **libq/20180504**

I am not the creator of the library, but I am using this library for a while now, mostly in combination with libcurl to interact with a backend webserver, and wanted to hopefully promote some interest in the library by providing this package. The original author does not update it anymore it seems, but it's quite stable. Because the author never really officially released a 1.0, but mentions that the api would not change anymore before 1.0, I have used the date of the commit from which I created this package instead of 1.0.

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by metalMajor** [comment]
> Hi! Thanks a lot for taking the time to contribute the recipe, we really appreciate it!
> 
> I've left some comments on the minor changes, but I'd suggest looking into the CMake package template https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package - basing the recipe off of this should be straightforward and will take care of some the pitfalls I've mentioned and most which I didn't :)
> 
> Let me know if you need any help with that

Hello, I think I managed to update it based on the template. Thank for the review!

**Comment by metalMajor** [comment]
@AbrilRBS and @uilianries : Can you live with this package now not being built on windows because otherwise I keep getting stuck on something in the build pipeline that nobody seems to be able to fix. Because I don't have Windows installed, only linux and macos, it is a really painful process to fix the Windows build.

**Comment by uilianries** [comment]
> @AbrilRBS and @uilianries : Can you live with this package now not being built on windows because otherwise I keep getting stuck on something in the build pipeline that nobody seems to be able to fix. Because I don't have Windows installed, only linux and macos, it is a really painful process to fix the Windows build.
> 
> In a later PR we can then fix Windows?

Totally! Just need to be clear in the validate() about that limitation.

File: recipes/libq/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    @property
+    def _build_tests(self):
+        return not self.conf.get("tools.build:skip_test", default=True, check_type=bool)
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def requirements(self):
+        if self._build_tests:
+            self.test_requires("gtest/1.14.0")
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:50]:
```suggestion
```

I would request removing testing requirements if possible. I see it required extra steps to make it work and even in the template there is no unit tests listed.

File: recipes/libq/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    def _patch_sources(self):
+        replace_in_file(self, os.path.join(self.source_folder, "CMakeLists.txt"),
+                        'include_directories( "3rdparty/gtest-1.8.0/include/" )',
+                        "")
+        replace_in_file(self, os.path.join(self.source_folder, "CMakeLists.txt"),
+                        'add_subdirectory( "3rdparty/gtest-1.8.0" )',
+                        "find_package(GTest REQUIRED)\n")
+
+    def build(self):
+        self._patch_sources()
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:83]:
```suggestion
    def build(self):
```

File: recipes/libq/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+        rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
+        rmdir(self, os.path.join(self.package_folder, "share"))
+        rm(self, "*.la", os.path.join(self.package_folder, "lib"))
+        rm(self, "*.pdb", os.path.join(self.package_folder, "lib"))
+        rm(self, "*.pdb", os.path.join(self.package_folder, "bin"))
+
+    def package_info(self):
+        self.cpp_info.libs = collect_libs(self)
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:101]:
```suggestion
        self.cpp_info.libs = ["q"]
```

Avoid using collect_libs, as we can not predict what libraries should be packaged.

File: recipes/libq/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        if self.settings.os in ["Windows"]:
+            raise ConanInvalidConfiguration(f"{self.ref} can not be built on Windows")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["BUILD_SHARED_LIBS"] = self.options.shared
+        tc.variables["q_BUILD_TESTS"] = self._build_tests
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:69]:
```suggestion
        tc.variables["q_BUILD_TESTS"] = False
```

File: recipes/libq/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            )
+        if self.settings.os in ["Windows"]:
+            raise ConanInvalidConfiguration(f"{self.ref} can not be built on Windows")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["BUILD_SHARED_LIBS"] = self.options.shared
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:68]:
```suggestion
        tc.variables["BUILD_SHARED_LIBS"] = self.options.shared
```

Is BUILD_SHARED_LIBS really needed? The CMakeToolchain should manage it already: https://docs.conan.io/2/reference/tools/cmake/cmaketoolchain.html#extending-and-advanced-customization

File: recipes/libq/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    def validate(self):
+        if self.settings.compiler.cppstd:
+            check_min_cppstd(self, self._min_cppstd)
+        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+        if minimum_version and Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
+            )
+        if self.settings.os in ["Windows"]:
+            raise ConanInvalidConfiguration(f"{self.ref} can not be built on Windows")
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:61]:
```suggestion
        if self.settings.os == "Windows":
            # FIXME: Test package is not capable to find q.lib
            raise ConanInvalidConfiguration(f"{self.ref} Conan recipe is not supported in Windows. Contributions are welcome.")
```

File: recipes/libq/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    name = "libq"
+    description = "A platform-independent promise library for C++, implementing asynchronous continuations."
+    license = "Apache-2.0"
+    topics = ("async", "promises")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/grantila/q"
+    settings = "os", "compiler", "build_type", "arch"
+    package_type = "library"
+    options = {
+        "shared": [True, False]
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:23]:
```suggestion
        "shared": [True, False],
        "fPIC": [True, False]
```

No fPIC support?

**Review by uilianries — COMMENTED** [review]
Thank you for your contribution, I did some suggestions to your current recipe.

File: recipes/libq/all/conanfile.py (hunk 4 above)
Comment (metalMajor) [reply recipes/libq/all/conanfile.py]: applied

File: recipes/libq/all/conanfile.py (hunk 5 above)
Comment (metalMajor) [reply recipes/libq/all/conanfile.py]: it's indeed not needed, looking at other packages too you don't need to set it

**Comment by metalMajor** [comment]
Thank you for the review, I applied your changes, it looks good now I think 👍 
I was not logged in anymore somehow, so I did not see the buttons to apply your changes(which I found strange), therefore I just copied them, and afterwards found out that I was not logged in.

**Review by uilianries — DISMISSED** [review]
LGTM

**Comment by AbrilRBS** [comment]
Hi @metalMajor I made some changes to get Windows building at least as static.

Now the only remaing hurdle for this PR to solve comes from the CLA bot, https://github.com/conan-io/conan-center-index/pull/22828#issuecomment-1953876477 says that there is at least one commit authored by a different author, who would need to also sign it (Or for the commit to be rebased with a different author)

**Comment by uilianries** [comment]
The use `Tom Deblauwe`. Probably, a second email that was used to push this PR. That email needs to sign the CLA too.

**Comment by metalMajor** [comment]
I think it is solved, I removed an email address, now I put it back.

**Review by AbrilRBS — APPROVED** [review]
Thanks a lot for taking the time to get the PR thru the finish line, and thanks for your patience while we got it reviewed, we appreciate it :)

File: recipes/libq/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        cmake_layout(self, src_folder="src")
+
+    def export_sources(self):
+        export_conandata_patches(self)
+
+    def validate(self):
+        check_min_cppstd(self, 11)
+        if self.settings.os == "Windows" and self.options.shared:
+            # It exports no symbols, so it can't be used as a shared library
+            raise ConanInvalidConfiguration("Does not support shared libraries on Windows")
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:44]:
```suggestion
	def config_options(self):
	    if self.settings.os == "Windows":
	        del self.options.fPIC
	        self.package_type = "static-library"

    def validate(self):
        check_min_cppstd(self, 11)        
```

I would suggest enforcing Windows libs as static only.

**Review by uilianries — DISMISSED** [review]
LGTM

File: recipes/libq/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
+import os
+
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.build import check_min_cppstd
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.files import copy, get, rm, rmdir, export_conandata_patches, apply_conandata_patches
+from conan.tools.microsoft import is_msvc
+from conan.tools.scm import Version
```
Comment (uilianries) [inline recipes/libq/all/conanfile.py:9]:
```suggestion
from conan.tools.build import check_min_cppstd
from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
from conan.tools.files import copy, get, rm, rmdir, export_conandata_patches, apply_conandata_patches
from conan.tools.microsoft import is_msvc
```

Keep linter clean

**Review by AbrilRBS — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM. Thank you for your PR!

**Comment by metalMajor** [comment]
Awesome!

**Pull request merged**